Add in-gameplay rank display indicating minimum/maximum achievable rank

normalid-awa opened the pull request:
I've added a feature to the rank display component that allows displaying the maximum and minimum ranks. It works similar to the accuracy display, where we have a maximum and minimum value shown for accuracy. I implemented this feature in the rank display, so now it can show both the minimum-achievable and maximum-achievable ranks.

In order to enable the rank display component to utilize this functionality, I have added two new properties, `MinimumRank` and `MaximumRank`, to the `ScoreProcessor` class.

sorry for the bad English :(

Preview: 

https://github.com/ppy/osu/assets/52519933/703f2ebe-be98-41d4-afdc-613e51c2d631

https://github.com/ppy/osu/assets/52519933/03b8ab61-8a50-40c4-8057-f21f761f1276

##### Even though the new Osu! update has been implementing the rank display update but I think it's good to implement the max. and min. achievable rank display. Just like how the accuracy did.It allows players to actually see what rank they can if they mess up at the following or what rank they can achieve if they get all perfectly.

normalid-awa committed:
Implemented features

bdach commented:
Would you mind explaining what is going on with https://github.com/ppy/osu/pull/28583 getting closed and then this getting opened?

jkh675 commented:
> Would you mind explaining what is going on with #28583 getting closed and then this getting opened?

I'm sorry for that, my repo master branch get out of sync and I don't know how to fix it. So that I deleted my repo, but I don't know it will cause the pr getting close, so I copy all the change and make a new pr

jkh675 commented:
It seems that my master branch is currently ahead of the osu! branch. But I accidentally merged my feature branch into the master branch, and it appears that the action to undo was not explicitly outlined in the documentation. sorry again for causing confusion.

normalid-awa committed:
Merge branch 'master' into features/minimum-and-maximum-rank-display

bdach commented on osu.Game/Rulesets/Scoring/ScoreProcessor.cs:
```diff
@@ -98,6 +98,16 @@ public partial class ScoreProcessor : JudgementProcessor
 
         private readonly Bindable<ScoreRank> rank = new Bindable<ScoreRank>(ScoreRank.X);
 
+        /// <remarks>
+        /// Using the minimum accuracy as the data source to calculate the minimum rank.
+        /// </remarks>
+        public readonly Bindable<ScoreRank> MinimumRank = new Bindable<ScoreRank>(ScoreRank.X);
+
+        /// <remarks>
+        /// Using the maximum accuracy as the data source to calculate the minimum rank.
+        /// </remarks>
+        public readonly Bindable<ScoreRank> MaximumRank = new Bindable<ScoreRank>(ScoreRank.X);
```
I'm having very mixed feelings about this living in `ScoreProcessor`. Maximum and minimum accuracy I can live with, since it isn't possible to calculate that from outside, but the rank is derived entirely from the score and the hit statistics, so I'm not sure I want this to live here (rather than locally in the skinnable component).

normalid-awa replied:
> I'm having very mixed feelings about this living in `ScoreProcessor`. Maximum and minimum accuracy I can live with, since it isn't possible to calculate that from outside, but the rank is derived entirely from the score and the hit statistics, so I'm not sure I want this to live here (rather than locally in the skinnable component).

From my perspective, I think it would be suitable for the minimum and maximum rank to be calculated within the `ScoreProcessor`. It appears that the logic for determining the minimum and maximum rank is shared across different rulesets, and there are two components (`LegacyRankDisplay` and `DefaultRankDisplay`) that require bind to these data.

smoogipoo replied:
I probably wouldn't want `Accuracy` or `Score` or any of the other bindables to be publicly editable either. Not sure if there's a use for them to be.

I think I'm okay with this PR for the time being, but I'd like to see them exposed as `IBindable` instead to follow the shape of `Rank`. Reason being here is that mods have modified `Rank` directly in the past which is how we ended up with the current `IBindable`.

normalid-awa replied:
> I probably wouldn't want `Accuracy` or `Score` or any of the other bindables to be publicly editable either. Not sure if there's a use for them to be.
> 
> I think I'm okay with this PR for the time being, but I'd like to see them exposed as `IBindable` instead to follow the shape of `Rank`. Reason being here is that mods have modified `Rank` directly in the past which is how we ended up with the current `IBindable`.

Implement the change at the lastest commit, the `MaximumRank` and `MinimumRank` are now `IBindable<ScoreRank>`

bdach commented on osu.Game/Rulesets/Scoring/ScoreProcessor.cs:
```diff
@@ -382,11 +392,28 @@ private void updateRank()
+            ScoreRank newMaxRank = RankFromScore(MaximumAccuracy.Value, ScoreResultCounts);
+            ScoreRank newMinRank = RankFromScore(MinimumAccuracy.Value, ScoreResultCounts);
+
+            // Check if the misses are >= 1, when the misses are more then 1, it clamp max rank no more then A.
+            // Check Accuracy.Value >= accuracy_cutoff_s is for the reduce unnecessary assign causing the icon flicking
+            ScoreResultCounts.TryGetValue(HitResult.Miss, out int missesCount);
+            if (missesCount > 1 && Accuracy.Value >= accuracy_cutoff_s)
+                newMaxRank = ScoreRank.A;
+            // detect is 100% acc, so it earn SS or SS+ rank
+            else if (Accuracy.Value == 1.0d)
+                newMaxRank = newRank;
```
Why is all of this being done here?

Miss count bigger than 1 only prevents gaining S rank in osu! and taiko. catch and mania only look at the raw accuracy. So putting any ruleset-specific logic here is both unnecessary (because `RankFromScore` should already be handling this), and incorrect (because you're applying rules to rulesets that don't apply to them normally)

And what is that about "the icon flicking" too? The bindable debounces updates (by checking if the value actually changed). As long as you're not assigning to `{Minimum,Maximum}Rank.Value` every time, no "flicking" should be taking place.

bdach commented on osu.Game/Screens/Play/HUD/DefaultRankDisplay.cs:
```diff
@@ -37,9 +44,66 @@ protected override void LoadComplete()
+            {
+                switch (mode.OldValue)
+                {
+                    case RankDisplayMode.Standard:
+                        scoreProcessor.Rank.UnbindBindings();
+                        break;
+
+                    case RankDisplayMode.MinimumAchievable:
+                        scoreProcessor.MinimumRank.UnbindBindings();
+                        break;
+
+                    case RankDisplayMode.MaximumAchievable:
+                        scoreProcessor.MaximumRank.UnbindBindings();
+                        break;
```
This cannot be done. Calling `.UnbindBindings()` unilaterally on a public bindable like that will screw up other consumers.

Instead, create a new bindable local to this component, and bind it to / unbind it from `ScoreProcessor` bindables as required. (That is, if those continue to exist even.)

normalid-awa committed:
Merge branch 'master' into features/minimum-and-maximum-rank-display

normalid-awa committed:
Optimized code

normalid-awa committed:
Merge branch 'master' into features/minimum-and-maximum-rank-display

normalid-awa committed:
Improve code quality

normalid-awa commented:
```diff C#
+        private Bindable<ScoreRank> rankBinding = new Bindable<ScoreRank>();
```
this binding was added to the rank display component, as they handle the rank display update.

normalid-awa committed:
Merge branch 'master' into features/minimum-and-maximum-rank-display

normalid-awa committed:
Merge branch 'master' into features/minimum-and-maximum-rank-display

normalid-awa committed:
Update rank when revert replay, so that it wont show the wrong previous data

normalid-awa committed:
Merge branch 'features/minimum-and-maximum-rank-display' of https://github.com/jkh675/osu into features/minimum-and-maximum-rank-display

normalid-awa committed:
Fix the delay update of rank

normalid-awa committed:
Merge branch 'master' into features/minimum-and-maximum-rank-display

bdach commented on osu.Game/Rulesets/Scoring/ScoreProcessor.cs:
```diff
@@ -259,6 +269,7 @@ protected sealed override void ApplyResultInternal(JudgementResult result)
 
                 updateScore();
             }
+            updateRank();
```
At the point all of these extra `updateRank()` calls are being added, it sure *looks* like `updateRank()` and `updateScore()` should just be one method.

In fact if I squint at master code, I believe there may be at least one potential bug related to a missing `updateRank()` call:

https://github.com/ppy/osu/blob/63136311b585052304fd53e74c526efd9881a093/osu.Game/Rulesets/Scoring/ScoreProcessor.cs#L495

The theory goes that if this call happens to not change accuracy but *also* require a rank update, the second one will not happen.

normalid-awa replied:
if it's possible to just move the `updateRank()` calls in to `updateScore()` and rename it to  `updateRankAndScore()`? so that the `updateRank` will be execute right after the score update.

bdach replied:
Yes that is pretty much what I was proposing. Call it `updateScore()`, which kinda already encompasses rank.

normalid-awa replied:
Implement the change at the latest commit

normalid-awa committed:
Combine both updateRank and updateScore method

normalid-awa committed:
Reduce the unnecessary rank update

normalid-awa committed:
Accuracy change can trigger rank update

normalid-awa committed:
Add test and optimize the performance

normalid-awa committed:
fix code quality

normalid-awa committed:
Make the min, max rank property to IBindable, so that they can't modify from outside

normalid-awa committed:
Merge branch 'master' into features/minimum-and-maximum-rank-display

frenzibyte committed:
Split rank display mode logic to abstract class and use standard namings

bdach commented on osu.Game/Screens/Play/HUD/GameplayRankDisplay.cs:
```diff
@@ -0,0 +1,69 @@
+
+namespace osu.Game.Screens.Play.HUD
+{
+    public abstract partial class GameplayRankDisplay : CompositeDrawable, IHasCurrentValue<ScoreRank>
```
hesitant about this existing, the duplication was inoffensive before and the logic was straightforward, with this abstract class existing people may get dumb ideas

frenzibyte replied:
Can you elaborate further? What dumb ideas would people have from sharing code between default/legacy pieces in an abstract component?

bdach replied:
i'm generally bearish on abstract classes these days and prefer duplication if it's duplication of trivial code. later monstrosities like `ScoreTable` get created that cannot be set right without a day wasted on righting it

normalid-awa replied:
If it was in the other situation like there are only 10 lines of code that were duplicated, I would prefer to just not extract them, but if there are huge amounts of business logic duplications, then I think should probably extract them to a separate class, or function, not only for the better readability of code but also for the better maintenance for the further changes.

frenzibyte reviewed (commented):
The rest is fine, all previous concerns look to be correctly resolved now.

bdach commented on osu.Game/Rulesets/Scoring/ScoreProcessor.cs:
```diff
@@ -382,11 +400,19 @@ private void updateRank()
 
             ScoreRank newRank = RankFromScore(Accuracy.Value, ScoreResultCounts);
+            ScoreRank newMaxRank = RankFromScore(MaximumAccuracy.Value, ScoreResultCounts);
+            ScoreRank newMinRank = RankFromScore(MinimumAccuracy.Value, ScoreResultCounts);
```
worst case is worst case and should assume accordingly. the hit result dictionary that is passed to calculate minimum rank should be filling unjudged objects with misses

frenzibyte replied:
@normalid-awa will leave this one to you.

normalid-awa replied:
> @normalid-awa will leave this one to you.

fixed in commit [normalid-awa:features/minimum-and-maximum-rank-display#1061982](https://github.com/ppy/osu/pull/28614/commits/10619827d7fd044fa7183c9980a16e7fbcaf1c16)

normalid-awa committed:
Fix the bug of minimum rank get over S

frenzibyte commented on osu.Game/Rulesets/Scoring/ScoreProcessor.cs:
```diff
@@ -401,8 +401,21 @@ private void updateRank()
             ScoreRank newRank = RankFromScore(Accuracy.Value, ScoreResultCounts);
             ScoreRank newMaxRank = RankFromScore(MaximumAccuracy.Value, ScoreResultCounts);
+
+            //To prevent the minimum rank got into >S, before completed the entire map,
+            //So that in here it fill the rest of the part with miss.
+            // (i.e. simulating the worse situation of all misses)
+            //ref: https://github.com/ppy/osu/pull/28614#discussion_r1670392366
+            if (ScoreResultCounts.TryGetValue(HitResult.Miss, out int currentMissCounts))
+                ScoreResultCounts[HitResult.Miss] = MaxHits - JudgedHits + currentMissCounts;
+            else
+                ScoreResultCounts[HitResult.Miss] = MaxHits - JudgedHits;
```
Any reason not to create a temporary dictionary for this? This is too weird to read as-is.

Also remove the link to the discussion thread, there's no need for it. If anything, one can blame the comment and reach back to this PR.

normalid-awa replied:
I think creating a temporary dictionary is kinda expensive as any update will trigger this update. And this is syncronize, so i do think it won't effect the outputs

frenzibyte replied:
Then create it once and re-use it.

normalid-awa replied:
Implemented in [normalid-awa:features/minimum-and-maximum-rank-display#f8a8668](https://github.com/ppy/osu/pull/28614/commits/f8a8668ac0a2323b09a41ccaf83ee5988741ebe3)

normalid-awa committed:
Remove the link from comments

normalid-awa committed:
Refactor previous commit

frenzibyte commented on osu.Game/Rulesets/Scoring/ScoreProcessor.cs:
```diff
@@ -209,6 +213,7 @@ public ScoreProcessor(Ruleset ruleset)
         {
             Ruleset = ruleset;
 
+            minimumResultCounts[HitResult.Miss] = MaxHits;
```
I think you're meant to put this in `ScoreProcessor.Reset`

normalid-awa replied:
Implemented in [ebede5c](https://github.com/ppy/osu/pull/28614/commits/ebede5c79d2ae137708f1bbce8be715d64804115)
Merge state

normalid-awa committed:
Move the code of initialize the minimum result count to reset method

frenzibyte commented:
I get notifications for any change you make, there's no need to request review, it's adding more noise to my end.

normalid-awa commented:
> I get notifications for any change you make, there's no need to request review, it's adding more noise to my end.

sorry for that, I didn't know the changes can trigger the notification

frenzibyte committed:
Rename dictionary and use in a single place

It's made to work for a single specific case, it shouldn't be scattered
all over the class.

frenzibyte committed:
Add test coverage

frenzibyte commented on osu.Game/Rulesets/Scoring/ScoreProcessor.cs:
```diff
@@ -375,18 +393,34 @@ private void updateScore()
             ScoreRank newRank = RankFromScore(Accuracy.Value, ScoreResultCounts);
+            ScoreRank newMaxRank = RankFromScore(MaximumAccuracy.Value, ScoreResultCounts);
+
+            // this arbitrarily only fills HitResult.Miss because it's the only hit result that can affect the state of RankFromScore in the main rulesets (see OsuScoreProcessor).
+            // eventually (whenever it's needed), other miss result types should be correctly filled and this dictionary should be updated regularly next to ScoreResultCounts,
+            // but this is sufficient for now.
+            minimumAccuracyResultCounts[HitResult.Miss] = MaxHits - JudgedHits + ScoreResultCounts.GetValueOrDefault(HitResult.Miss);
```
@ppy/team-client requesting review for the addition of this part in particular to resolve https://github.com/ppy/osu/pull/28614#discussion_r1670392366

bdach replied:
Too hacky to live imo. The remaining results should be populated so that this is in a _sane_ state.

Also I'm not even convinced this is right - what if some of the remaining judgements are `IgnoreMiss`, for example?

frenzibyte replied:
Partly addressed in https://github.com/ppy/osu/pull/28614/commits/471d7f2aedf97800c7b2bd1367855524f9bcd7c1, see https://github.com/ppy/osu/pull/28614#issuecomment-2436363284.

normalid-awa committed:
Merge branch 'master' into features/minimum-and-maximum-rank-display

frenzibyte commented:
Please do not merge master branch, it adds unnecessary noise to our notification inbox.

bdach commented on osu.Game.Rulesets.Osu.Tests/TestSceneOsuScoreProcessor.cs:
```diff
@@ -0,0 +1,86 @@
+    public partial class TestSceneOsuScoreProcessor : OsuTestScene
+    {
+        [Test]
+        public void TestMinimumRankNeverReachesSRank()
```
This should read "never reaches S rank before completion". That should have test coverage too.

normalid-awa replied:
Maybe should make the minimum rank calculations separate into a virtual method, and can be overwritten when needed, (e.g. the situation that misses or others can affect the rank)

normalid-awa committed:
Rename the test `TestMinimumRankNeverReachesSRank` to `TestMinimumRankShouldNeverReachSRankBeforeComplete`

frenzibyte committed:
Merge branch 'master' into features/minimum-and-maximum-rank-display

frenzibyte committed:
Move "minimum accuracy miss" logic to `OsuScoreProcessor`

frenzibyte committed:
Ensure minimum rank can be S/SS on completion

frenzibyte commented:
I've took another round on this PR, moving the special logic of minimum rank to `OsuScoreProcessor` and simplifying it there. While it's still wrong as pointed out by https://github.com/ppy/osu/pull/28614#discussion_r1686117423, I don't think it's that noticable since it only matters for the very last seconds of the beatmap where all that's left is the tail of a slider or the bonus ticks of a spinner, and I've made sure that the rank still promotes to S/SS when the player completes the beatmap and there's an outro playing.

If that's still not convincing enough to move this PR forward then I'll mark it off my inbox.

peppy commented on osu.Game.Rulesets.Osu/Scoring/OsuScoreProcessor.cs:
```diff
@@ -17,14 +17,21 @@ public OsuScoreProcessor()
+            => adjustRank(base.RankFromScore(accuracy, results), results.GetValueOrDefault(HitResult.Miss));
+
+        protected override ScoreRank MinimumRankFromScore(double accuracy, IReadOnlyDictionary<HitResult, int> results)
+            // this will be wrong when the remaining judgements do not affect accuracy and the player did not miss before,
+            // but it's a bit of a tiny detail to fix. we'll at least show an S/SS rank when the user completes the beatmap
+            // (especially when the beatmap has an outro storyboard).
+            => adjustRank(base.RankFromScore(accuracy, results), JudgedHits == MaxHits ? results.GetValueOrDefault(HitResult.Miss) : int.MaxValue);
```
This comment says that "what" but not the "why". If I read this in isolation (which is the first thing I did when seeing the history of this PR), my first question is going to be "why will it be wrong?".

I also don't really follow why the special case logic can't remain in `RankFromScore` as it was before this PR. If i dig deeper I'm sure I can understand why, but it's not really passing the smell test at this point for sane code. Which is probably not great since these changes are being made in `ScoreProcessor`, not some isolated skinnable UI element class..

frenzibyte replied:
This comment is written specifically to explain the "why" part: "why will it be wrong? because *when the remaining judgements do not affect accuracy and the player did not miss before* then the rank will be incorrectly A rather than S/SS".

This is `OsuScoreProcessor` so I would like to imagine there's a level of locality to this change. However, I can move this to become a workaround from a UI element point, sounds like a better plan to keep the `ScoreProcessor` class clean and consistent from changes like these.

frenzibyte replied:
After further discussion, I understand where the concern comes from. The comment is meant to highlight the `JudgedHits == MaxHits` particularly, but it looks like it's trying to explain why this entire override exists. The comment is also written without setting up the scenario exactly, beginning from "it's wrong" but not...explaining why it's wrong.

Take a look at https://github.com/ppy/osu/pull/28614/commits/53272c16a0ca22a7dde9e7378c1199c9c1f6716c and see if it helps.

frenzibyte committed:
Rename method and better explain logic
